### PR TITLE
[SPARK-17447] Performance improvement in Partitioner.defaultPartitioner without sortBy

### PR DESCRIPTION
**Pull request by codlife**
## What changes were proposed in this pull request?

if there are many rdds in some situations,the sort will loss he performance servely,actually we needn't sort the rdds , we can just scan the rdds one time to gain the same goal.
## How was this patch tested?

manual tests

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/Partitioner.scala [hunk 1]
```diff
@@ -55,14 +55,15 @@ object Partitioner {
    * We use two method parameters (rdd, others) to enforce callers passing at least 1 RDD.
    */
   def defaultPartitioner(rdd: RDD[_], others: RDD[_]*): Partitioner = {
-    val bySize = (Seq(rdd) ++ others).sortBy(_.partitions.length).reverse
-    for (r <- bySize if r.partitioner.isDefined && r.partitioner.get.numPartitions > 0) {
-      return r.partitioner.get
+    val rdds = Seq(rdd) ++ others
+    val filteredRdds = rdds.filter( _.partitioner.exists(_.numPartitions > 0 ))
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/Partitioner.scala]: This doesn't match the code I posted in minor ways. There should be no extra spaces around operators; there should be a space after 'if'; this uses `return` unnecessarily. This is a lot of time spent on a trivial change, so I'd appreciate it if you read the guidance on things like style, and read feedback carefully if you're pursuing this. Or else close this.

Comment (codlife) [inline core/src/main/scala/org/apache/spark/Partitioner.scala]: @srowen thank you very much , i am a new hand about saprk, but i'm interested in it very much,i have fixed my code style.thanks.

File: core/src/main/scala/org/apache/spark/Partitioner.scala [hunk 2]
```diff
@@ -55,14 +55,16 @@ object Partitioner {
...
    */
   def defaultPartitioner(rdd: RDD[_], others: RDD[_]*): Partitioner = {
-    val bySize = (Seq(rdd) ++ others).sortBy(_.partitions.length).reverse
-    for (r <- bySize if r.partitioner.isDefined && r.partitioner.get.numPartitions > 0) {
-      return r.partitioner.get
-    }
-    if (rdd.context.conf.contains("spark.default.parallelism")) {
-      new HashPartitioner(rdd.context.defaultParallelism)
+    val rdds = (Seq(rdd) ++ others)
+    val hashPartitioner = rdds.filter(_.partitioner.exists(_.numPartitions > 0))
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/Partitioner.scala]: hasPartitioner, not hashPartitioner. You should copy the code I provided.

Comment (codlife) [inline core/src/main/scala/org/apache/spark/Partitioner.scala]: @srowen  Thank you ,I will lean much about code style.

Comment (codlife) [inline core/src/main/scala/org/apache/spark/Partitioner.scala]: First time to commit, but enjoy the process,i have updated.

Comment (codlife) [inline core/src/main/scala/org/apache/spark/Partitioner.scala]: @srowen Please check, Thank you  very much!

**Comment by srowen** [comment]
Jenkins test this please

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**